app/fetch-did-web-doc.py

```diff
@@ -0,0 +1,145 @@
+    response = request(url)
+
+    result = json.dumps(
+        {
+            "url": url,
+            "nonce": nonce,
+            "status_code": response["status_code"],
+            "body": base64.b64encode(response["body"]).decode("utf-8"),
+        }
+    ).encode("utf-8")
+
+    return result
+
+
+def fetch_attested(url, nonce):
+    success = False
+    retries = HTTP_RETRIES
+    with tempfile.NamedTemporaryFile() as out_path:
+        args = [
+            f"{AFETCH_DIR}/afetch",
+            f"{AFETCH_DIR}/libafetch.enclave.so.signed",
+            out_path.name,
+            url,
+            nonce,
+        ]
+        logging.info(f"Starting {' '.join(args)}")
+        while not success and retries >= 0:
+            retries -= 1
+            try:
+                subprocess.run(
+                    args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, check=True
+                )
+                success = True
+                break
+            except subprocess.CalledProcessError as e:
+                logging.error(f"afetch failed with code {e.returncode}: {e.stdout}")
+            except Exception as e:
+                logging.error(f"Unknown error: {e}")
+            if retries >= 0:
+                logging.info(f"Retrying in {HTTP_DEFAULT_RETRY_AFTER} seconds")
+                time.sleep(HTTP_DEFAULT_RETRY_AFTER)
+            else:
+                raise e
+        if success:
+            with open(out_path.name, "rb") as f:
+                result = f.read()
+            logging.info(f"afetch succeeded, output size is {len(result)} bytes")
+
+    return result
+
+
+def request(url, data=None, headers=None):
+    logging.info(f"Requesting {url}")
+    if headers is None:
+        headers = {}
+    req = Request(url, data, headers)
+    # Transport-level security is not required here as the content
+    # sent back to CCF is signed by the attested-fetch enclave through
+    # hardware attestation. Because of that, we can safely disable
+    # certificate validation and don't have to deal with CCF's self-signed certs.
+    ctx = ssl.create_default_context()
+    ctx.check_hostname = False
+    ctx.verify_mode = ssl.CERT_NONE
+    retries = HTTP_RETRIES
+    success = False
+    status_code = 0
+    body = b""
+    while not success and retries >= 0:
+        retries -= 1
+        try:
+            response = urlopen(req, context=ctx, timeout=CONNECT_TIMEOUT)
+            status_code = response.getcode()
+            body = response.read()
+            success = True
+            logging.info(f"HTTP status {status_code}")
+            break
+        except HTTPError as e:
+            status_code = e.getcode()
+            body = e.read()
+            retry_after = int(e.headers.get("Retry-After", HTTP_DEFAULT_RETRY_AFTER))
+            logging.error(f"HTTP status {status_code}: {body}")
+            if status_code not in [429, 503]:
+                break
+        except URLError as e:
```

Comment:
Should we be recording those errors and returning them to the ledger? Stuff like unresolvable hostnames or unable to establish a connection seems useful.

Reply:
I think recording that there was a network error is fine, though I think using the special status code 0 for that is enough. Note that this piece of code is only for the unattested case, and we'll do the same as part of #105 for attested. I think going from no error reporting to HTTP errors and whether there was a network error is good enough for now.